File: src/main/java/net/thenamedev/legendapi/message/Message.java
package net.thenamedev.legendapi.message;

import net.thenamedev.legendapi.utils.Rank;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;

/**
 * Created on 5/24/2015
 *
 * @author ThePixelDev
 */
public class Message {

    private MessageType type;
    private StringBuilder builder;
    private int[] times = {1, 4, 1};

    public Message() {
        type = MessageType.CHAT;
        builder = new StringBuilder();
    }

    public Message(MessageType type) {
        this.type = type;
        builder = new StringBuilder();
    }

    public Message append(String append) {
        builder.append(append);
        return this;
    }

    public void send(Player... players) {
        String built = toString();
        for(Player p : players) {
            switch(type) {
                case CHAT:
                    p.sendMessage(built);
                    break;
                case ACTIONBAR:
                    MessageUtilities.sendActionbar(p, built);
                    break;
                case SUBTITLE:
                    MessageUtilities.sendSubtitle(p, built, times[0], times[1], times[2]);
                    break;
                case TITLE:
                    MessageUtilities.sendTitle(p, built, times[0], times[1], times[2]);
                    break;
            }
        }
    }

    public void broadcast() {
        for(Player p : Bukkit.getOnlinePlayers()) {
            send(p);
        }
    }

    public String toString() {
        return builder.toString();
    }

    public void setDisplayTime(int in, int stay, int out) {
        if(!(type == MessageType.TITLE || type == MessageType.SUBTITLE))
            throw new ClassCastException("Cannot set display time for chat or action bar messages");
        times[0] = in;
        times[1] = stay;
        times[2] = out;
    }

    public int[] getDisplayTimes() {
        return times.clone();
    }

    public void broadcast(Rank r) {
        for(Player p : Bukkit.getOnlinePlayers()) {
            if(!Rank.isRanked(p, r))
                continue;
            send(p);
        }
    }
}

File: src/old/main/java/legendapi/message/Message.kt
package legendapi.message

import legendapi.fanciful.FancyMessage
import legendapi.log.BukLog
import legendapi.utils.Rank
import legendarena.LegendArena
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.Sound
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player

class Message {

    private var type = MessageType.CHAT

    private var fanciful: FancyMessage? = null

    private var fadeIn = 5
    private var stay = 10
    private var fadeOut = 5

    private var sound: Sound? = null
    private var pitch = 1f

    private var builder: StringBuilder? = null

    private var reflector = MessageReflector()

    public constructor() {}

    public constructor(type: MessageType) {
        this.type = type
    }

    public fun append(msg: String): Message {
        if(type == MessageType.FANCIFUL)
            throw ClassCastException("You can't append a String onto a FancyMessage!")
        if(builder == null)
            builder = StringBuilder()
        builder!!.append(msg)
        return this
    }

    public fun append(color: ChatColor): Message {
        if(type == MessageType.FANCIFUL)
            throw ClassCastException("You can't append a ChatColor onto a FancyMessage!")
        if(builder == null)
            builder = StringBuilder()
        builder!!.append("" + color)
        return this
    }

    public fun append(msg: FancyMessage): Message {
        if(type != MessageType.FANCIFUL) {
            if(builder != null)
                builder = null
        }
        this.fanciful = msg
        return this
    }

    public fun setSound(s: Sound): Message {
        sound = s
        return this
    }

    public fun setPitch(p: Float): Message {
        pitch = p
        return this
    }

    public fun send(vararg p: Player) {
        var msg = toString()

        for(p1 in p) {
            if(sound != null)
                p1.playSound(p1.getLocation(), sound, pitch, pitch)
            if(msg != "")
                _send(p1)
        }
    }

    public fun send(p: CommandSender) {
        if(sound != null)
            (p as Player).playSound(p.getLocation(), sound, pitch, pitch)
        _send(p as Player)
    }

    private fun _send(p: Player) {
        var msg = toString()
        if((msg == "" || msg == "\n") && sound != null) return
        when(type) {
            MessageType.CHAT -> p.sendMessage(msg)
            MessageType.ACTIONBAR -> sendActionbar(p)
            MessageType.TITLE -> sendTitle(p)
            MessageType.SUBTITLE -> sendSubtitle(p)
            MessageType.FANCIFUL -> fanciful!!.send(p)
        }
    }

    public fun broadcast() {
        for(p in Bukkit.getOnlinePlayers())
            send(p)
    }

    public fun broadcast(r: Rank) {
        for(p in Bukkit.getOnlinePlayers()) {
            if(!r.isRanked(p)) continue
            send(p)
        }
    }

    /**
     * Even I have no idea why this function exists.
     */
    public fun broadcast(vararg r: Rank) {
        for(p in Bukkit.getOnlinePlayers())
            for(b in r) {
                if(!b.isRanked(p)) continue
                send(p)
                break
            }
    }

    public fun networkBroadcast() {
        throw UnsupportedOperationException("Soon[tm]")
    }

    override public fun toString(): String {
        if(builder == null)
            return ""
        return builder.toString()
    }

    private fun sendTitle(p: Player) {
        reflector.send(0, p, toString(), fadeIn, stay, fadeOut)
    }

    private fun sendSubtitle(p: Player) {
        reflector.send(1, p, toString(), fadeIn, stay, fadeOut)
    }

    private fun sendActionbar(p: Player) {
        reflector.sendActionbar(p, toString())
    }

}